### ISSUE TITLE
Support early stopping of XGboost

### ISSUE THREAD

**Issue by yutannihilation** [tidymodels/parsnip#237]
Though `early_stopping_rounds` argument of `xgb.train()` is accessible on `set_engine()`, it also requires `watchlist`, a named list of `xgb.DMatrix` datasets to use for evaluating model performance and it's a bit hard to create this. This is because how parsnip converts `data.frame` to `xgb.DMatrix` internally is not exposed to us. For this reason, I needed to peek at the code inside parsnip and create this function:

``` r
create_xgb_dmatrix <- function(d) {
  o <- parsnip:::convert_form_to_xy_fit(y ~ ., d, composition = "matrix")
  xgboost::xgb.DMatrix(data = o$x, label = as.integer(o$y == "yes"), missing = NA)
}

watchlist <- list(
  train = create_xgb_dmatrix(analysis(fold))),
  test = create_xgb_dmatrix(assessment(fold)))
)
```

Is it possible to add some interface for early stopping? Or can `convert_form_to_xy_fit()` be exposed eventually?

**Comment by lorenzwalthert** [comment]
IMHO, the missing early stopping feature really is why a lot of people still use the xgboost package directly instead of mlr or parsnip, so I think it would be really great to have support for it. The line of argument basically goes "xgboost is the best single algorithm for tabular data and you get rid of a hyper parameter when you use early stopping so it saves a lot of time".

Reference: https://github.com/mlr-org/mlr/issues/1398#event-2915471655

**Comment by juliasilge** [comment]
There has been other interest in having exported `convert_form_to_xy_fit()`, such as #86 . Looks like we should consider that, aside from truly supporting early stopping more fully.

**Comment by cbrioso** [comment]
> Though `early_stopping_rounds` argument of `xgb.train()` is accessible on `set_engine()`, it also requires `watchlist`, a named list of `xgb.DMatrix` datasets to use for evaluating model performance and it's a bit hard to create this. This is because how parsnip converts `data.frame` to `xgb.DMatrix` internally is not exposed to us. For this reason, I needed to peek at the code inside parsnip and create this function:
> 
> ```r
> create_xgb_dmatrix <- function(d) {
>   o <- parsnip:::convert_form_to_xy_fit(y ~ ., d, composition = "matrix")
>   xgboost::xgb.DMatrix(data = o$x, label = as.integer(o$y == "yes"), missing = NA)
> }
> 
> watchlist <- list(
>   train = create_xgb_dmatrix(analysis(fold))),
>   test = create_xgb_dmatrix(assessment(fold)))
> )
> ```
> 
> Is it possible to add some interface for early stopping? Or can `convert_form_to_xy_fit()` be exposed eventually?

I'm not clear how you can add early stopping time on set_engine. Could you please elaborate?

**Comment by yutannihilation** [comment]
OK, in case the lengthy example is useful, here it is:
https://gist.github.com/yutannihilation/d2569bf3fbef05403979a8e2171b1880

(Disclaimer: I'm still a beginner of machine learning things, so I'm not sure the parameters are appropriate...)

**Comment by topepo** [comment]
I have some prototype code. I just have not had the time to test and include it yet. It's on my to-do list.

**Comment by cbrioso** [comment]
> OK, in case the lengthy example is useful, here it is:
> https://gist.github.com/yutannihilation/d2569bf3fbef05403979a8e2171b1880
> 
> (Disclaimer: I'm still a beginner of machine learning things, so I'm not sure the parameters are appropriate...)

Fantastic. Thanks!

**Comment by topepo** [comment]
Keep in mind that the implementation that I'm doing would use a randomly sampled validation set from the data passed in for modeling (as opposed to passing in a watchlist with a data set there). 

The rationale for this is that the data may have been pre-processed and we couldn't guarantee that the data in the watchlist would be encoded in the same was as the data given to the underlying function.

**Comment by topepo** [comment]
Implemented in 

```r
# requires tidyr >= 1.0.0
devtools::install_github("tidymodels/parsnip@early-stop")
devtools::install_github("tidymodels/dials@stop-iter")
```

There is a main parameter for `boost_tree(stop_iter)`. For now, you'll also have to use a new version of `dials` if you want to use `tune_*1`. 

The low level function wrapper: 

``` r
library(tidymodels)
#> ── Attaching packages ───────────────────────────────────────────────────── tidymodels 0.1.0 ──
#> ✓ broom     0.5.4          ✓ recipes   0.1.12    
#> ✓ dials     0.0.6.9000     ✓ rsample   0.0.6     
#> ✓ dplyr     0.8.5          ✓ tibble    3.0.1     
#> ✓ ggplot2   3.3.0          ✓ tune      0.1.0     
#> ✓ infer     0.5.1          ✓ workflows 0.1.0     
#> ✓ parsnip   0.1.1.9000     ✓ yardstick 0.0.5     
#> ✓ purrr     0.3.4
#> Warning: package 'rsample' was built under R version 3.6.2
#> ── Conflicts ──────────────────────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard()  masks scales::discard()
#> x dplyr::filter()   masks stats::filter()
#> x dplyr::lag()      masks stats::lag()
#> x ggplot2::margin() masks dials::margin()
#> x recipes::step()   masks stats::step()
library(modeldata                                                                                                                  )
library(doMC)
#> Loading required package: foreach
#> 
#> Attaching package: 'foreach'
#> The following objects are masked from 'package:purrr':
#> 
#>     accumulate, when
#> Loading required package: iterators
#> Loading required package: parallel
registerDoMC(cores = 10)

data(cells, package = "modeldata")

cells <- cells %>% select(-case)

set.seed(114313)
xgb_train(x = cells[, -1], y = cells$class, nrounds = 500)
#> ##### xgb.Booster
#> raw: 400.9 Kb 
#> call:
#>   xgboost::xgb.train(params = list(eta = 0.3, max_depth = 6, gamma = 0, 
#>     colsample_bytree = 1, min_child_weight = 1, subsample = 1), 
#>     data = x, nrounds = 500, watchlist = wlist, verbose = 0, 
#>     objective = "binary:logistic")
#> params (as set within xgb.train):
#>   eta = "0.3", max_depth = "6", gamma = "0", colsample_bytree = "1", min_child_weight = "1", subsample = "1", objective = "binary:logistic", silent = "1"
#> xgb.attributes:
#>   niter
#> callbacks:
#>   cb.evaluation.log()
#> # of features: 56 
#> niter: 500
#> nfeatures : 56 
#> evaluation_log:
#>     iter training_error
#>        1       0.131253
#>        2       0.108965
#> ---                    
#>      499       0.000000
#>      500       0.000000

set.seed(114313)
xgb_train(x = cells[, -1], y = cells$class, nrounds = 500, validation = .2)
#> ##### xgb.Booster
#> raw: 172.1 Kb 
#> call:
#>   xgboost::xgb.train(params = list(eta = 0.3, max_depth = 6, gamma = 0, 
#>     colsample_bytree = 1, min_child_weight = 1, subsample = 1), 
#>     data = x, nrounds = 500, watchlist = wlist, verbose = 0, 
#>     objective = "binary:logistic")
#> params (as set within xgb.train):
#>   eta = "0.3", max_depth = "6", gamma = "0", colsample_bytree = "1", min_child_weight = "1", subsample = "1", objective = "binary:logistic", silent = "1"
#> xgb.attributes:
#>   niter
#> callbacks:
#>   cb.evaluation.log()
#> # of features: 56 
#> niter: 500
#> nfeatures : 56 
#> evaluation_log:
#>     iter validation_error
#>        1         0.217957
#>        2         0.193808
#> ---                      
#>      499         0.196285
#>      500         0.196904

set.seed(114313)
xgb_train(x = cells[, -1], y = cells$class, nrounds = 500, validation = .2, 
          early_stop = 10)
#> ##### xgb.Booster
#> raw: 25.8 Kb 
#> call:
#>   xgboost::xgb.train(params = list(eta = 0.3, max_depth = 6, gamma = 0, 
#>     colsample_bytree = 1, min_child_weight = 1, subsample = 1), 
#>     data = x, nrounds = 500, watchlist = wlist, verbose = 0, 
#>     early_stopping_rounds = 10, objective = "binary:logistic")
#> params (as set within xgb.train):
#>   eta = "0.3", max_depth = "6", gamma = "0", colsample_bytree = "1", min_child_weight = "1", subsample = "1", objective = "binary:logistic", silent = "1"
#> xgb.attributes:
#>   best_iteration, best_msg, best_ntreelimit, best_score, niter
#> callbacks:
#>   cb.evaluation.log()
#>   cb.early.stop(stopping_rounds = early_stopping_rounds, maximize = maximize, 
#>     verbose = verbose)
#> # of features: 56 
#> niter: 19
#> best_iteration : 9 
#> best_ntreelimit : 9 
#> best_score : 0.191331 
#> nfeatures : 56 
#> evaluation_log:
#>     iter validation_error
#>        1         0.217957
#>        2         0.193808
#> ---                      
#>       18         0.197523
#>       19         0.195046

set.seed(114313)
xgb_train(x = cells[, -1], y = cells$class, nrounds = 500, validation = .2, 
          early_stop = 10, eval_metric = "auc")
#> ##### xgb.Booster
#> raw: 39 Kb 
#> call:
#>   xgboost::xgb.train(params = list(eta = 0.3, max_depth = 6, gamma = 0, 
#>     colsample_bytree = 1, min_child_weight = 1, subsample = 1), 
#>     data = x, nrounds = 500, watchlist = wlist, verbose = 0, 
#>     early_stopping_rounds = 10, objective = "binary:logistic", 
#>     eval_metric = "auc")
#> params (as set within xgb.train):
#>   eta = "0.3", max_depth = "6", gamma = "0", colsample_bytree = "1", min_child_weight = "1", subsample = "1", objective = "binary:logistic", eval_metric = "auc", silent = "1"
#> xgb.attributes:
#>   best_iteration, best_msg, best_ntreelimit, best_score, niter
#> callbacks:
#>   cb.evaluation.log()
#>   cb.early.stop(stopping_rounds = early_stopping_rounds, maximize = maximize, 
#>     verbose = verbose)
#> # of features: 56 
#> niter: 37
#> best_iteration : 27 
#> best_ntreelimit : 27 
#> best_score : 0.87968 
#> nfeatures : 56 
#> evaluation_log:
#>     iter validation_auc
#>        1       0.829858
#>        2       0.841892
#> ---                    
#>       36       0.879025
#>       37       0.878357

set.seed(114313)
xgb_train(x = cells[, -1], y = cells$class, nrounds = 500, 
          early_stop = 10, eval_metric = "auc")
#> ##### xgb.Booster
#> raw: 90 Kb 
#> call:
#>   xgboost::xgb.train(params = list(eta = 0.3, max_depth = 6, gamma = 0, 
#>     colsample_bytree = 1, min_child_weight = 1, subsample = 1), 
#>     data = x, nrounds = 500, watchlist = wlist, verbose = 0, 
#>     early_stopping_rounds = 10, objective = "binary:logistic", 
#>     eval_metric = "auc")
#> params (as set within xgb.train):
#>   eta = "0.3", max_depth = "6", gamma = "0", colsample_bytree = "1", min_child_weight = "1", subsample = "1", objective = "binary:logistic", eval_metric = "auc", silent = "1"
#> xgb.attributes:
#>   best_iteration, best_msg, best_ntreelimit, best_score, niter
#> callbacks:
#>   cb.evaluation.log()
#>   cb.early.stop(stopping_rounds = early_stopping_rounds, maximize = maximize, 
#>     verbose = verbose)
#> # of features: 56 
#> niter: 39
#> best_iteration : 29 
#> best_ntreelimit : 29 
#> best_score : 1 
#> nfeatures : 56 
#> evaluation_log:
#>     iter training_auc
#>        1     0.922023
#>        2     0.950475
#> ---                  
#>       38     1.000000
#>       39     1.000000
```

<sup>Created on 2020-05-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by cbrioso** [comment]
Marvelous, thanks.

**Comment by yutannihilation** [comment]
> Keep in mind that the implementation that I'm doing would use a randomly sampled validation set from the data passed in for modeling (as opposed to passing in a watchlist with a data set there).

Sounds good to me! I'm not familiar with the cross-validation workflow of XGBoost, but I bet it's rare to use different watchlists for the same dataset, so probably this is enough

**Comment by cbrioso** [comment]
Thanks Max

**Comment by lorenzwalthert** [comment]
Thanks for your work Max. I working with unbalanced data now and one thing I cannot do with `xgboost::xgb.cv()`, but with `xgboost::xgb.train()` is to upsample the training data and leave the validation data as is. This is necessary to do proper validation I think. With your approach described above, would it be possible to apply a recipes pre-processing step such as `themis::step_smote(..., skip = TRUE)` to the data after you split into train and validation? I wonder because of 
> The rationale for this is that the data may have been pre-processed and we couldn't guarantee that the data in the watchlist would be encoded in the same was as the data given to the underlying function.

**Comment by topepo** [comment]
Closed by #304
